numpy/core/src/umath/ufunc_object.c

```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
     PyObject *obj;
 
     /* Convert and fill in input arguments */
+    int all_scalar = 1;
+    int any_scalar = 0;
+    int all_legacy = 1;
     for (int i = 0; i < nin; i++) {
         obj = PyTuple_GET_ITEM(full_args.in, i);
 
         if (PyArray_Check(obj)) {
-            PyArrayObject *obj_a = (PyArrayObject *)obj;
-            out_op[i] = (PyArrayObject *)PyArray_FromArray(obj_a, NULL, 0);
+            out_op[i] = (PyArrayObject *)obj;
+            Py_INCREF(out_op[i]);
+        }
+        else {
+            /* Convert the input to an array and check for special cases */
+            out_op[i] = (PyArrayObject *)PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (out_op[i] == NULL) {
+                goto fail;
+            }
+        }
+        out_borrowed_op_DTypes[i] = NPY_DTYPE(PyArray_DESCR(out_op[i]));
+        if (!out_borrowed_op_DTypes[i]->legacy) {
+            all_legacy = 0;
+        }
+        if (PyArray_NDIM(out_op[i]) != 0) {
+            all_scalar = 0;
+            continue;
         }
         else {
-            out_op[i] = (PyArrayObject *)PyArray_FromAny(obj,
-                                    NULL, 0, 0, 0, NULL);
+            any_scalar = 1;
+        }
+        if ((PyObject *)out_op[i] == obj) {
+            continue;  /* input was an array: no need to check for pyscalar */
```

Comment:
See comment above. If not refactoring, I'd write this as != and put all the next type-checking statements inside.

Reply:
Adopted for now, but you are right, there may well be a better way to write this. Coming "back" to the original object is a bit strange.